### PR TITLE
Pipes - Eva - Hotel

### PR DESCRIPTION
**Pull request by Esawyer25**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |    I tried to have a date comparison class since so many other people were finding that useful and I knew my tests to check if the rooms were unreserved, unbooked, free for the entire range, ect. were cumbersome, seemed repetitive and relied on building a data_structure and not behavior.   But I could not figure out how to make use of such a class and eventually did without it.  I am interested to see how others made that work.  |
| Describe a concept that you gained more clarity on as you worked on this assignment. |   I didn't need to store all the information about all the reservations and blocks all the time, but just call them when I needed them.  |
| Describe a _nominal test_ that you wrote for this assignment. | Make_reservation can make a reservation for several days.   |
| Describe an _edge case test_ that you wrote for this assignment. | Make_reservation can make a reservation for just one day.   |
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I spent several hours writing psuedocode for the entire first-draft of wave 1 before writing any real code this time so I think I did well with that.  For the initial tests I think I did well writing tests before code, but later when I was moving methods from one class to another it got less regimented   |

### CODE REVIEW

File: lib/hotel.rb [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+      return nil
+    end
+
+    #can be used for "As an administrator, I can access the list of
+    # reservations for a specific date"
+    def find_reservations_by_date(date)
+      reservations_on_date = []
+      @all_reservations.each do |reservation|
+        if reservation.nights_booked.include?(date)
+          reservations_on_date << reservation
```
Comment (droberts-sea) [inline lib/hotel.rb]: This is an example of tight coupling between `Reservation` and `Hotel`.  This `Hotel` code reaches into `Reservation` and inspects its list of dates directly. A cleaner way to organize things might be to write a method `Reservation#contain?(date)`, and call that here. In that case, this method could be simplified to:

```ruby
return @all_reservations.select { |reservation| reservation.contain?(date) }
```

File: lib/hotel.rb [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+    # Their is one inner hash per night.
+    def unreserved_and_unblocked(nights)
+      if nights.class == Date
+        nights = [nights]
+      end
+      unblocked = find_all_unblocked_rooms(nights)
+      unreserved = find_all_unreserved_rooms(nights)
+      available_over_range = []
+      index = 0
+      unblocked.each do |night|
```
Comment (droberts-sea) [inline lib/hotel.rb]: While this code works, I suspect that you have overcomplicated things a bit. Here is a strategy that might be a little cleaner to implement:

```
Start with a list of all the rooms
For each reservation:
    If the reservation overlaps the given dates
        remove the reservation's room from the list
For each block:
    If the block overlaps the given dates
        remove all the block's rooms from the list
```

the resulting list will include only rooms that are not unavailable for those dates.

File: lib/hotel.rb [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+    #Can be used for "As an administrator, I can get the total cost
+    # for a given reservation" Or, you could just do reservation.cost
+    def get_cost(reservation)
+      reservation.cost
+    end
+
+    def date_range(first_night, last_night)
+      if first_night != last_night
+        @nights = (first_night..last_night)
+      else
```
Comment (droberts-sea) [inline lib/hotel.rb]: I'm not sure you want to use an instance variable here - that will set `@nights` for the entire `Hotel`. Instead, creating and returning a local variable is probably the right choice.

Comment (droberts-sea) [reply lib/hotel.rb]: This method would be a great place to do some error checking on the range of dates.

File: specs/coverage/.last_run.json [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
+  "result": {
```
Comment (droberts-sea) [inline specs/coverage/.last_run.json]: Your coverage file should be included in your `.gitignore`

File: specs/hotel_spec.rb [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+    end
+
+    it "can make a one-night reservation" do
+      @ritz.must_be_kind_of My_Hotel::Hotel
+      holiday = @ritz.make_reservation(@feb1)
+      holiday.must_be_kind_of My_Hotel::Reservation
+    end
+
+    it "raises an argument if you try to reserve a room and none is available" do
+      20.times do
```
Comment (droberts-sea) [inline specs/hotel_spec.rb]: Good failure case!

File: specs/hotel_spec.rb [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+      feb3_reservations.length.must_equal 2
+      feb6_reservations = @ritz.find_reservations_by_date(@feb6)
+      feb6_reservations.length.must_equal 1
+    end
+
+    it "given a date, it returns an empty array if there are no reservations on that date" do
+      @ritz.make_reservation(@feb1, @feb5)
+      may6_reservations = @ritz.find_reservations_by_date(@may6)
+      may6_reservations.must_be_empty
+    end
```
Comment (droberts-sea) [inline specs/hotel_spec.rb:107]: Again, good coverage, especially this second test.

File: specs/hotel_spec.rb [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+      @ritz.make_block(@feb1, @feb5, [1, 2, 3, 4], 0.75)
+      @ritz.make_reservation(@feb1, @feb3)
+      reservation_and_block = @ritz.unreserved_and_unblocked(@feb1)
+      reservation_and_block[0].length.must_equal 15
+      reservation_and_block = @ritz.unreserved_and_unblocked(@feb5)
+    end
+  end
+
+  describe "find_continious_open_room" do
+    it "must return an hash of room =>prices" do
```
Comment (droberts-sea) [inline specs/hotel_spec.rb]: Somewhere in these tests, you should have coverage around when a room counts as being  available. My entire list of cases includes:
- Not available:
  - Same dates
  - Overlaps in the front
  - Overlaps in the back
  - Completely contained
  - Completely containing
- Available:
  - Completely before
  - Completely after
  - Ends on the checkin date
  - Starts on the checkout date

**Comment by droberts-sea** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| yes - good work
Answer comprehension questions	| yes - for an example of using a `DateRange` class to simplify code and isolate tests, [check out our reference implementation](https://github.com/droberts-ada/hotel/tree/dpr/solution).
**Design** |
Each class is responsible for a single piece of the program | mostly - It seems like there's a lot of functionality contained within `Hotel` (more than one responsibility) that could reasonably be pushed into other classes (not enough responsibility). An example of this is checking if a reservation or block overlaps a date (or date range). This will also help make it easier to avoid tight coupling.
Classes are loosely coupled | mostly - see inline comment
**Wave 1** |
List rooms | yes
Reserve a room for a given date range | yes
List reservations for a given date | yes
Calculate reservation price | yes
Invalid date range produces an error | no
Test coverage | yes
**Wave 2** |
View available rooms for a given date range | yes
Reserving a room that is not available produces an error | yes
Test coverage | missing some - see inline comment
**Wave 3** |
Create a block of rooms | yes
Check if a block has rooms | yes
Reserve a room from a block | yes
Test coverage | yes
**Additional Feedback** |

Good job overall. Your high-level design seems deliberate and well-thought-out, though there are still a couple places where the `Hotel` class takes on responsibilities it probably shouldn't. There were a few spots where specific Ruby code ended up more complicated than it needed to be, but in general I'm happy with your logic, and quite impressed by your test coverage. Keep up the hard work!

**Comment by Esawyer25** [comment]
I think i responded to the (very helpful!) comments from the last pull request.  I also changed a few other minor things, like my Hotel#find_by_block_id and  Hotel#find_by_reservation_id.
